crates/wysiwyg/src/tests/test_to_markdown.rs

```diff
@@ -0,0 +1,228 @@
+    assert_to_md(
+        "abc <del>line1<br />line2<br /><br />line3</del> def",
+        r#"abc ~~line1\
+line2\
+\
+line3~~ def"#,
+    );
+}
+
+#[test]
+fn text_with_underline() {
+    assert_to_md_no_roundtrip("<u>abc</u>", "abc");
+}
+
+#[test]
+fn text_with_inline_code() {
+    assert_to_md("<code>abc</code>", "`` abc ``");
+    // Inline code with a backtick inside.
+    assert_to_md("<code>abc ` def</code>", "`` abc ` def ``");
+    // Inline code with a backtick at the start.
+    assert_to_md("<code>`abc</code>", "`` `abc ``");
+    assert_to_md("abc <code>def</code> ghi", "abc `` def `` ghi");
+    assert_to_md("abc<code> def </code>ghi", "abc``  def  ``ghi");
+    // It's impossible to get a line break inside an inline code with Markdown.
+    assert_to_md_no_roundtrip(
+        "abc <code>line1<br />line2<br /><br />line3</code> def",
+        "abc `` line1 line2  line3 `` def",
+    );
+    assert_to_md_no_roundtrip(
+        "abc <code>def <strong>ghi</strong> jkl</code> mno",
+        "abc `` def __ghi__ jkl `` mno",
+    );
+}
+
+#[test]
+fn link() {
+    assert_to_md(r#"<a href="url">abc</a>"#, "[abc](<url>)");
+    assert_to_md_no_roundtrip(r#"<a href="u<rl">abc</a>"#, r#"[abc](<u\<rl>)"#);
+    // Empty link.
+    assert_to_md(r#"<a href="">abc</a>"#, r#"[abc](<>)"#);
+    // Formatting inside link.
+    assert_to_md(
+        r#"<a href="url">abc <strong>def</strong> ghi</a>"#,
+        r#"[abc __def__ ghi](<url>)"#,
+    );
+    assert_to_md(r#"<a href="(url)">abc</a>"#, r#"[abc](<\(url\)>)"#);
+}
+
+#[test]
+fn list_unordered() {
+    assert_to_md(
+        r#"<ul><li>item1</li><li>item2</li></ul>"#,
+        r#"* item1
+* item2"#,
+    );
+
+    assert_to_md_no_roundtrip(
+        r#"<ul><li>item1<ul><li>subitem1</li><li>subitem2</li></ul></li><li>item2</li></ul>"#,
+        r#"* item1
+  * subitem1
+  * subitem2
+* item2"#,
+    );
+}
+
+#[test]
+fn list_ordered() {
+    assert_to_md(
+        r#"<ol><li>item1</li><li>item2</li></ol>"#,
+        r#"1. item1
+2. item2"#,
+    );
+}
+
+fn assert_to_md_no_roundtrip(html: &str, expected_markdown: &str) {
+    let markdown = to_markdown(html);
+    assert_eq!(markdown, expected_markdown);
+}
+
+fn assert_to_md(html: &str, expected_markdown: &str) {
+    let markdown = to_markdown(html);
+    assert_eq!(markdown, expected_markdown);
+
+    let expected_html = html;
+    let html = to_html(&markdown);
+
+    assert_eq!(html, expected_html);
+}
+fn to_markdown(html: &str) -> Utf16String {
+    let markdown = cm(&format!("{html}|")).state.dom.to_markdown();
+    assert!(markdown.is_ok());
+
+    markdown.unwrap()
+}
+
+fn to_html(markdown: &Utf16String) -> String {
+    use md_parser::{html, Options, Parser};
+
+    let mut options = Options::empty();
+    options.insert(Options::ENABLE_STRIKETHROUGH);
+
+    let markdown = markdown.as_ustr().to_string_lossy();
+
+    let parser = Parser::new_ext(&markdown, options);
+
+    let mut html = String::new();
```

Comment:
This confused me having a variable called `html` as well as `md_parser::html` so close together. Can we think of a different name for the variable?

Reply:
Sure.

Reply:
Done.